crates/red_knot_python_semantic/src/types/diagnostic.rs

```diff
@@ -1313,6 +1314,46 @@ pub(crate) fn report_invalid_arguments_to_annotated(
     ));
 }
 
+pub(crate) fn report_bad_argument_to_get_protocol_members(
+    context: &InferContext,
+    call: &ast::ExprCall,
+    class: ClassLiteralType,
+) {
+    let Some(builder) = context.report_lint(&INVALID_ARGUMENT_TYPE, call) else {
+        return;
+    };
+    let db = context.db();
+    let mut diagnostic = builder.into_diagnostic("Invalid argument to `get_protocol_members`");
+    diagnostic.set_primary_message("This call will raise `TypeError` at runtime");
+    diagnostic.info("Only protocol classes can be passed to `get_protocol_members`");
+
+    let class_scope = class.body_scope(db);
+    let class_node = class_scope.node(db).expect_class();
+    let class_name = &class_node.name;
+    let class_def_diagnostic_range = TextRange::new(
+        class_name.start(),
+        class_node
+            .arguments
+            .as_deref()
+            .map(Ranged::end)
+            .unwrap_or_else(|| class_name.end()),
+    );
+    let mut class_def_diagnostic = SubDiagnostic::new(
+        Severity::Info,
+        format_args!("`{class_name}` is declared here, but it is not a protocol class:"),
+    );
+    class_def_diagnostic.annotate(Annotation::primary(
+        Span::from(class_scope.file(db)).with_range(class_def_diagnostic_range),
+    ));
+    diagnostic.sub(class_def_diagnostic);
+
+    diagnostic.info(
+        "A class is only a protocol class if it directly inherits \
+            from `typing.Protocol` or `typing_extensions.Protocol`",
+    );
+    diagnostic.info("See https://typing.python.org/en/latest/spec/protocol.html#");
+}
```

Comment:
This is a lovely diagnostic, and a great example for what we can be doing with the new diagnostic framework in other cases -- but it's also a lot of code for an extremely narrow case (misuse of one specific function)! Not really suggesting we change anything here, just not sure it will be worth investing this much in every narrow edge-case diagnostic...

Reply:
Yeah, I know what you mean.

I think there might be ways to make the diagnostic APIs a little less verbose? The code here isn't that complicated, just a little boilerplate-y. And I think the way you get a reputation for having "excellent error messages" like rustc is by paying attention to lots of little details like this.